compiler/qsc_qasm3/src/semantic/lowerer.rs

```diff
@@ -834,25 +880,42 @@ impl Lowerer {
         let ty_span = stmt.ty.span();
         let name = stmt.identifier.name.clone();
         let qsharp_ty = self.convert_semantic_type_to_qsharp_type(&ty.clone(), stmt.ty.span());
-        let symbol = Symbol {
-            name: name.to_string(),
-            ty: ty.clone(),
-            qsharp_ty,
-            span: stmt.identifier.span,
-            io_kind: IOKind::Default,
+        let init_expr = match &stmt.init_expr {
+            syntax::ValueExpr::Expr(expr) => {
+                let expr = self.lower_expr(expr);
+                self.cast_expr_with_target_type_or_default(Some(expr), &ty, stmt.span)
+            }
+            syntax::ValueExpr::Measurement(measure_expr) => self.lower_measure_expr(measure_expr),
         };
 
-        // process the symbol and init_expr gathering any errors
-        let init_expr = self.lower_expr_with_target_type(Some(&stmt.init_expr), &ty, stmt.span);
+        let mut symbol = Symbol::new(
+            &name,
+            stmt.identifier.span,
+            ty.clone(),
+            qsharp_ty,
+            IOKind::Default,
+        );
+
+        if init_expr.ty.is_const() {
+            symbol = symbol.with_const_expr(Rc::new(init_expr.clone()));
+        }
 
         let symbol_id =
             self.try_insert_or_get_existing_symbol_id(name, symbol, stmt.identifier.span);
 
+        if !init_expr.ty.is_const() {
+            self.push_semantic_error(SemanticErrorKind::ExprMustBeConst(
+                "const decl init expr".to_string(),
+                init_expr.span,
+            ));
+            return semantic::StmtKind::Err;
+        }
```

Comment:
I don't think we should return stmtkind::err here. We've pushed a semantic error which will prevent complation, and we've been able to lower all of the pieces, we just have a ty error.